### DISCUSSION TITLE
Does AbstractLoginFormAuthenticator implements onAuthenticationFailure method wrongly?

### DISCUSSION THREAD

**Discussion by lecneri** [symfony/symfony#46547] [Q&A] [answered]
The class `Symfony\Component\Security\Http\Authenticator\AbstractLoginFormAuthenticator` extends `AbstractAuthenticator` in same namespace. The `AbstractAuthenticator` implements `AuthenticatorInterface` which defines the following method

```php
public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response;
```
But the method in `AbstractLoginFormAuthenticator` is defined as follows:
```php
public function onAuthenticationFailure(Request $request, AuthenticationException $exception): Response;
```

I belive that should be `?Response` instead of `Response` only.

Also I don't know how it's not throwing an error because these methods signatures doesn't match.

**Comment by mbabker** [comment, marked as answer]
Because of type covariance, a child implementation can have a stricter return type than its parent, so long as what it returns fulfills the requirements of the parent.  The `AbstractLoginFormAuthenticator::onAuthenticationFailure()` method always returns a `Response` instance, which satisfies the interface's requirement that the method returns either a `Response` or null.

See https://www.php.net/manual/en/language.oop5.variance.php for more info on how this works.
